### PR TITLE
Windows Containerd Pre-submits

### LINKED ISSUE #1857: Add PR support for the Windows containerd solution
/kind feature

**Describe the solution you'd like**
[A clear and concise description of what you want to happen.]
The [dev cluster template](https://github.com/kubernetes-sigs/cluster-api-provider-azure/tree/main/templates/test/dev) has ability to deploy windows nodes but is missing the part to use the custom binaries like we have for the dockershim solution: https://github.com/kubernetes-sigs/cluster-api-provider-azure/blob/feba723784818f210a7c4cb96e82d7b38d0b2334/templates/test/dev/cluster-template-custom-builds-windows.yaml#L420-L429

**Anything else you would like to add:**
[Miscellaneous information that will assist in solving the issue.]

**Environment:**

- cluster-api-provider-azure version: 
- Kubernetes version: (use `kubectl version`): 
- OS (e.g. from `/etc/os-release`): 

/sig windows

### PR DESCRIPTION
**Pull request by jsturtevant**
**What type of PR is this?**

/kind feature

**What this PR does / why we need it**:

Enabled containerd Pre-submit tests for windows

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #1857 

**Special notes for your reviewer**:

_Please confirm that if this PR changes any image versions, then that's the sole change this PR makes._

**TODOs**:

- [ ] squashed commits
- [ ] includes documentation
- [ ] adds unit tests

**Release note**:

```release-note
NONE
```

/sig windows

### CODE REVIEW

**Comment by jsturtevant** [comment]
The PR jobs for Linux and Windows passed on the PR I set up here for testing https://github.com/kubernetes/kubernetes/pull/109032


/assign @marosset @CecileRobertMichon

**Comment by jsturtevant** [comment]
Opened: https://github.com/kubernetes-sigs/cluster-api-provider-azure/issues/2197 for the test failure which was not related to changes here 

/retest

File: templates/test/dev/cluster-template-custom-builds.yaml [hunk 1]
```diff
@@ -375,6 +384,39 @@ spec:
           New-Item -ItemType Directory -Force -Path C:\tmp\
         path: C:/create-temp-folder.ps1
         permissions: "0744"
+      - content: |
```
Comment (marosset) [inline templates/test/dev/cluster-template-custom-builds.yaml:387]: Can we set `$ErrorActionPreference Stop` at the beginning of this script?
I'm worried that if we don't and there is an issue with any of the packages they might get skipped over and we won't test what we think we are testing.

File: templates/test/dev/cluster-template-custom-builds.yaml [hunk 2]
```diff
@@ -375,6 +384,39 @@ spec:
...
+          # Tag it to the ci version.  The image knows how to use the copy locally with the configmap
+          # that is applied at at this stage (windows-kubeproxy-ci.yaml)
+          ctr.exe -n k8s.io images pull docker.io/sigwindowstools/kube-proxy:v1.23.1-calico-hostprocess
+          ctr.exe -n k8s.io images tag docker.io/sigwindowstools/kube-proxy:v1.23.1-calico-hostprocess "docker.io/sigwindowstools/kube-proxy:${CI_VERSION/+/_}-calico-hostprocess"
+
+          kubeadm.exe version -o=short
+          kubectl.exe version --client=true --short=true
+          kubelet.exe --version
+          kube-proxy.exe --version
+        path: C:/replace-k8s-binaries.ps1
```
Comment (marosset) [inline templates/test/dev/cluster-template-custom-builds.yaml:418]: replace-node-binaries.ps1?
This is replacing containerd too.

File: templates/test/dev/cluster-template-custom-builds.yaml (hunk 1 above)
Comment (marosset) [reply templates/test/dev/cluster-template-custom-builds.yaml:387]: We should do this everywhere TBH

File: templates/test/dev/cluster-template-custom-builds.yaml [hunk 3]
```diff
@@ -375,6 +384,39 @@ spec:
...
         path: C:/create-temp-folder.ps1
         permissions: "0744"
+      - content: |
+          Stop-Service kubelet -Force
+
+          $$binaries=@("kubeadm", "kubectl", "kubelet", "kube-proxy")
+          $$ci_url="https://${AZURE_STORAGE_ACCOUNT}.blob.core.windows.net/${JOB_NAME}/${KUBE_GIT_VERSION}/bin/windows/amd64"
+          foreach ( $$binary in $$binaries )
+          {
+            echo "installing package: $$binary ${KUBE_GIT_VERSION}"
```
Comment (marosset) [inline templates/test/dev/cluster-template-custom-builds.yaml:394]:
```suggestion
            echo "installing binary: $$binary ${KUBE_GIT_VERSION}"
```

File: templates/test/dev/cluster-template-custom-builds.yaml [hunk 4]
```diff
@@ -375,6 +384,39 @@ spec:
...
+            curl.exe --retry 10 --retry-delay 5 "$$ci_url/$$binary.exe" --output "c:/k/$$binary.exe"
+          }
+
+          $$CONTAINERD_URL="${WINDOWS_CONTAINERD_URL}"
+          if($$CONTAINERD_URL -ne ""){
+            Stop-Service containerd -Force
+            echo "downloading containerd: $$CONTAINERD_URL"
+            curl.exe --retry 10 --retry-delay 5 -L "$$CONTAINERD_URL" --output "c:/k/containerd.tar.gz"
+            cmd /c tar -zxvf c:/k/containerd.tar.gz -C "c:/Program Files/containerd" --strip-components 1
```
Comment (marosset) [inline templates/test/dev/cluster-template-custom-builds.yaml:403]:
```suggestion
            tar.exe -zxvf c:/k/containerd.tar.gz -C "c:/Program Files/containerd" --strip-components 1
```

Just calling tar.exe should have the same effect as calling `tar` from cmd w/o starting extra processes

Comment (marosset) [reply templates/test/dev/cluster-template-custom-builds.yaml:403]: similar to curl.exe

File: templates/test/dev/cluster-template-custom-builds.yaml [hunk 5]
```diff
@@ -375,6 +384,39 @@ spec:
...
+            containerd.exe --version
```
Comment (marosset) [inline templates/test/dev/cluster-template-custom-builds.yaml:404]: Is there a reason to not output containerd hcsshim version always?
I think that info might be helpful (especially the shim version) which isn't available w/ `kubectl get node`

File: templates/test/dev/cluster-template-custom-builds.yaml (hunk 2 above)
Comment (jsturtevant) [reply templates/test/dev/cluster-template-custom-builds.yaml:418]: It was mimicking the same scripts on linux, and initially didn't replace the containerd versions

File: templates/test/dev/cluster-template-custom-builds.yaml (hunk 5 above)
Comment (jsturtevant) [reply templates/test/dev/cluster-template-custom-builds.yaml:404]: The binary doesn't output a version as of now: https://github.com/microsoft/hcsshim/issues/1268

Comment (marosset) [reply templates/test/dev/cluster-template-custom-builds.yaml:404]: So should line 405 be commented it out?
It is currently trying to run `containerd-shim-runhcs-v1.exe --version`

My initial comment in this thread was meant to suggest moving all of the --version calls outside of the conditional.

File: templates/test/ci/prow-ci-version/patches/kubeadm-bootstrap-windows.yaml [hunk 6]
```diff
@@ -2,6 +2,8 @@
   path: /spec/template/spec/files/-
```
Comment (marosset) [inline templates/test/ci/prow-ci-version/patches/kubeadm-bootstrap-windows.yaml:2]: unrelated to this PR but would folks like to update this repo so the generated templates show up as generated files in PRs?
I can open up a PR to do so if people want. 
@CecileRobertMichon @jsturtevant

File: templates/test/dev/cluster-template-custom-builds.yaml (hunk 5 above)
Comment (jsturtevant) [reply templates/test/dev/cluster-template-custom-builds.yaml]: updated

File: templates/test/dev/cluster-template-custom-builds.yaml (hunk 2 above)
Comment (jsturtevant) [reply templates/test/dev/cluster-template-custom-builds.yaml]: updated to use `replace-ci-binaries`

**Comment by marosset** [comment]
/lgtm

Thanks for the updates!

**Comment by jsturtevant** [comment]
both known flakes
/retest

**Comment by jsturtevant** [comment]
both Linux and Windows tests passed on the stub PR in k/k: https://github.com/kubernetes/kubernetes/pull/109032. 

/assign @CecileRobertMichon @mboersma

**Review by CecileRobertMichon — COMMENTED** [review]
/approve

**Pull request merged**